hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -187,15 +262,15 @@ object InferPType {
         InferPType(body, env.bind(accumName -> zero.pType2, valueName -> a.pType2.asInstanceOf[PArray].elementType))
         assert(body.pType2 == zero.pType2)
 
-        coerce[PStreamable](a.pType2).copyStreamable(zero.pType2)
+        coerce[PStreamable](a.pType2).copyStreamable(zero.pType2, zero.pType2.required)
```

Comment:
Similiarly here, the requiredness of `ArrayScan` should be the same as the requiredness of `a`; you'll need to unify the requirednesses of `zero` and `body` to get the correct element type.

Reply:
Addressed. Here zero.pType2 and body.pType2 are required to be equal, so their requiredeness is required to be the same.

Reply:
Requiring them to be equal will currently invalidate totally valid IRs, though. (This was a latent bug and part of the reason we needed ptypes to deal with unifying requiredness, I'm pretty sure!)